### PR TITLE
Add aslr for secondary stack

### PR DESCRIPTION
**Pull request by renlord**

### CODE REVIEW

**Comment by renlord** [comment]
addresses https://github.com/GrapheneOS/os_issue_tracker/issues/11 and https://github.com/GrapheneOS/os_issue_tracker/issues/12

File: libc/bionic/pthread_create.cpp [hunk 1]
```diff
@@ -220,14 +220,25 @@ ThreadMapping __allocate_thread_mapping(size_t stack_size, size_t stack_guard_si
...
+#else
+  size_t max_gap_size = stack_size / 10;
+#endif
+  size_t gap_size = __BIONIC_ALIGN(arc4random_uniform(max_gap_size), PAGE_SIZE);
+
+  // Allocate in order: stack guard, stack, (random) guard page(s), pthread_internal_t, static TLS, guard page.
   size_t mmap_size;
   if (__builtin_add_overflow(stack_size, stack_guard_size, &mmap_size)) return {};
-  if (__builtin_add_overflow(mmap_size, PTHREAD_GUARD_SIZE, &mmap_size)) return {};
+  // Make sure random stack top guard size are multiples of PAGE_SIZE.
```
Comment (thestinger) [inline libc/bionic/pthread_create.cpp]: This comment seems to be in the wrong place.

File: libc/bionic/pthread_create.cpp [hunk 2]
```diff
@@ -307,6 +326,7 @@ static int __allocate_thread(pthread_attr_t* attr, bionic_tcb** tcbp, void** chi
     memset(thread, 0, sizeof(pthread_internal_t));
   }
 
+
```
Comment (thestinger) [inline libc/bionic/pthread_create.cpp]: Can delete this extra added line.

File: libc/include/sys/cdefs.h [hunk 3]
```diff
@@ -57,6 +57,8 @@
 #endif
 
 #define __BIONIC_ALIGN(__value, __alignment) (((__value) + (__alignment)-1) & ~((__alignment)-1))
+#define __BIONIC_ALIGN_DOWN(value, alignment) ((value) & ~((alignment) - 1))
+
```
Comment (thestinger) [inline libc/include/sys/cdefs.h:61]: Also don't need to add a line here.

File: libc/bionic/pthread_create.cpp [hunk 4]
```diff
@@ -295,7 +314,7 @@ static int __allocate_thread(pthread_attr_t* attr, bionic_tcb** tcbp, void** chi
   } else {
     mapping = __allocate_thread_mapping(0, PTHREAD_GUARD_SIZE);
     if (mapping.mmap_base == nullptr) return EAGAIN;
-
+    // custom mapping contains random guard gap between stack and pthread_internal_t.
```
Comment (thestinger) [inline libc/bionic/pthread_create.cpp]: This comment can be left out.

**Review by thestinger — CHANGES_REQUESTED** [review]

File: libc/bionic/pthread_create.cpp (hunk 4 above)
Comment (renlord) [reply libc/bionic/pthread_create.cpp]: fixed.

File: libc/include/sys/cdefs.h (hunk 3 above)
Comment (renlord) [reply libc/include/sys/cdefs.h:61]: fixed

File: libc/bionic/pthread_create.cpp (hunk 2 above)
Comment (renlord) [reply libc/bionic/pthread_create.cpp]: fixed

File: libc/bionic/pthread_create.cpp (hunk 1 above)
Comment (renlord) [reply libc/bionic/pthread_create.cpp]: fixed.

**Pull request merged**